### PR TITLE
Update docker container script for use with OpenJ9 builds

### PR DESCRIPTION
**Pull request by jdekonin**
* update jenkins container build for centos6.10
* update repos to alternate now that os is end of life
* update namespace to reference new repo location
* move perms adjusting to end of dockerfile so all content is owned by $user
* comment out freemarker requirement as its no longer in use
* remove non-existent ius-release-el6.rpm
* modified SDK location as builds were modified to reference $USER_HOME/bootjdks/

Signed-off-by: Joe deKoning <joe_dekoning@ca.ibm.com>

### CODE REVIEW

**Comment by jdekonin** [comment]
fyi @AdamBrousseau @vsebe

**Comment by jdekonin** [comment]
The change in location for freemarker.jar and bootjdks is to match expected locations for the automated builds.

I triggered a container build here (pointing at my repo):
https://ci.eclipse.org/openj9/view/Build/job/Build-Jenkins-Agent-Container/106/console

**Comment by genie-openj9** [comment]
Can one of the admins verify this patch?

**Comment by jdekonin** [comment]
The namespace needs to change still but there is a bugzilla  opened to address the fact the dockerhub repo isn't rw for the openj9 bot.

File: buildenv/docker/mkdocker.sh [hunk 1]
```diff
@@ -532,32 +543,33 @@ prepare_user() {
...
+  echo "RUN useradd -ms /bin/bash --uid $userid $user \\"
   echo " && mkdir /home/$user/.ssh \\"
   echo " && chmod 700 /home/$user/.ssh"
   echo "COPY authorized_keys known_hosts /home/$user/.ssh/"
-  echo "RUN chown -R $user:$user /home/$user \\"
-  echo " && chmod 600 /home/$user/.ssh/authorized_keys /home/$user/.ssh/known_hosts"
+  echo "RUN chmod 600 /home/$user/.ssh/authorized_keys /home/$user/.ssh/known_hosts \\"
+  echo " && echo \"PS1=\\\"\\u:DOCKER-CONTAINER $ \\\"\" > /home/$user/.bashrc \\"
+  echo " && echo \"ulimit -u 32000\" >> /home/$user/.bashrc \\"
+  echo " && echo \"ulimit -n 2048\" >> /home/$user/.bashrc"
```
Comment (keithc-ca) [inline buildenv/docker/mkdocker.sh:553]: Why do we care about the shell prompt string?
Much escaping can be avoided by using single quotes (`'`); the `$` should be escaped, though.
The two limits can be changed in one command: `ulimit -n 2048 -u 32000`.

File: buildenv/jenkins/docker-nodes/jenkinsfile-build-container.groovy [hunk 2]
```diff
@@ -42,9 +42,9 @@ if ("${OS}" == "unknown" || !(validArch.contains(ARCH))) {
 }
 
 def BUILD_OPTS = ""
-if ("${OS}" == "centos6.9") {
+if ("${OS}" == "centos6.10") {
     if ("${ARCH}" == "x86") {
-        BUILD_OPTS = "--dist=centos --version=6.9"
+        BUILD_OPTS = "--dist=centos --version=6.10 --cuda"
```
Comment (keithc-ca) [inline buildenv/jenkins/docker-nodes/jenkinsfile-build-container.groovy:47]: CUDA should be selected for ppc64le and x86, independent of the O/S.

File: buildenv/jenkins/docker-nodes/jenkinsfile-build-container.groovy [hunk 3]
```diff
@@ -62,7 +62,7 @@ if ("${BUILD_OPTS}" == "") {
 }
 
 def NODE = (params.NODE) ? params.NODE : "sw.tool.docker && hw.arch.${ARCH}"
-def NAMESPACE = (params.NAMESPACE) ? params.NAMESPACE : "eclipse-openj9"
+def NAMESPACE = (params.NAMESPACE) ? params.NAMESPACE : "openj9"
```
Comment (keithc-ca) [inline buildenv/jenkins/docker-nodes/jenkinsfile-build-container.groovy:65]: This is the namespace at dockerhub, right?

File: buildenv/docker/mkdocker.sh [hunk 4]
```diff
@@ -219,12 +220,22 @@ if [ $cuda != no ] ; then
...
+fi
 }
 
 install_centos_packages() {
+if [ $version = 6.10 ] ; then
+  echo "RUN sed -i -e 's|mirrorlist|#mirrorlist|' \\"
+  echo "           -e 's|#baseurl=http://mirror.centos.org/centos/\$releasever|baseurl=https://vault.centos.org/6.10|' \\"
+  echo "           /etc/yum.repos.d/CentOS-Base.repo"
+  echo ""
+fi
```
Comment (keithc-ca) [inline buildenv/docker/mkdocker.sh:236]: Why is this needed? Don't the default repo configurations work?

File: buildenv/docker/mkdocker.sh [hunk 5]
```diff
@@ -532,32 +543,33 @@ prepare_user() {
...
+  echo "RUN chown -R $user:$user /home/$user"
 }
 
 install_freemarker() {
   echo ""
   local freemarker_version=2.3.8
-  echo "# Download and extract freemarker.jar to /root/freemarker.jar."
-  echo "RUN mkdir -p /root \\"
-  echo " && cd /root \\"
+  echo "# Download and extract freemarker.jar to /home/$user/freemarker.jar."
```
Comment (keithc-ca) [inline buildenv/docker/mkdocker.sh:563]: Does `buildenv/ansible/roles/openj9-from-sources/tasks/main.yml` need to change too?

File: buildenv/jenkins/docker-nodes/jenkinsfile-build-container.groovy (hunk 3 above)
Comment (pshipton) [reply buildenv/jenkins/docker-nodes/jenkinsfile-build-container.groovy:65]: eclipseopenj9, not openj9? 
https://hub.docker.com/r/eclipseopenj9/jenkins-agent-x86-centos6

File: buildenv/docker/mkdocker.sh [hunk 6]
```diff
@@ -219,12 +220,22 @@ if [ $cuda != no ] ; then
   echo "FROM nvidia/cuda:${cuda}-devel-ubuntu16.04 AS cuda-dev"
   echo ""
 fi
+if [ $dist = centos ] ; then
+  echo "FROM $dist:$dist$version"
+else
   echo "FROM $dist:$version"
+fi
```
[the remark this answers is not in the document]
Comment (jdekonin) [reply buildenv/docker/mkdocker.sh:227]: When I initially did the tag change and repo change was what I needed to get this functional.  I'll try a run with this and the repo reverted.

File: buildenv/docker/mkdocker.sh (hunk 4 above)
Comment (jdekonin) [reply buildenv/docker/mkdocker.sh:236]: When I initially did the tag change and repo change was what I needed to get this functional.  I'll try a run with this and the repo reverted.

File: buildenv/docker/mkdocker.sh (hunk 1 above)
Comment (jdekonin) [reply buildenv/docker/mkdocker.sh:553]: I'll play with modifying.  The reason for the shell prompt was so there was an obvious eyecatcher to indicate that the docker container launch worked and that the job is compiling in the container and not the host.  Purely a debugging/confirmation bit of output for the user.

File: buildenv/docker/mkdocker.sh (hunk 5 above)
Comment (jdekonin) [reply buildenv/docker/mkdocker.sh:563]: I don't see how that is related to this bit of code?  But I suppose that can be modified as well.  The docker files that were replaced with this script were originally written for the build system to use which uses a basic user and not root user.  The user specific changes from /root -> /home/$user/ will mean a minor change to devs that use these yes, but if they are running as root they should have full access /home/$user/.

For freemarker.jar...is this actually a requirement for x/p/z linux platforms any more?

File: buildenv/jenkins/docker-nodes/jenkinsfile-build-container.groovy (hunk 2 above)
Comment (jdekonin) [reply buildenv/jenkins/docker-nodes/jenkinsfile-build-container.groovy:47]: I'll work in for all platforms.

File: buildenv/jenkins/docker-nodes/jenkinsfile-build-container.groovy (hunk 3 above)
Comment (jdekonin) [reply buildenv/jenkins/docker-nodes/jenkinsfile-build-container.groovy:65]: Correct, I hadn't had the update from the bugzilla I opened to create that repo until earlier today.  I will correct this as well.

File: buildenv/docker/mkdocker.sh (hunk 5 above)
Comment (keithc-ca) [reply buildenv/docker/mkdocker.sh:563]: x/p/z Linux all use cmake by default. It may be time to remove freemarker from the script.
Opinions, @pshipton?

Comment (pshipton) [reply buildenv/docker/mkdocker.sh:563]: How about we comment it out, so it's not installed by default, but a user can uncomment if needed.

Comment (keithc-ca) [reply buildenv/docker/mkdocker.sh:563]: If you mean hide the call to `install_freemarker` like this, I agree:
```
# install_freemarker
```

Comment (pshipton) [reply buildenv/docker/mkdocker.sh:563]: Yes, that is what I mean.

File: buildenv/docker/mkdocker.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (jdekonin) [reply buildenv/docker/mkdocker.sh]: Not required and removed from my next push.

File: buildenv/docker/mkdocker.sh (hunk 4 above)
Comment (jdekonin) [reply buildenv/docker/mkdocker.sh]: CentOS 6 is no longer hosted at mirror.centos.org as per http://mirror.centos.org/centos/6/readme so I modify the system default to redirected to the vault/archive location.

File: buildenv/docker/mkdocker.sh (hunk 1 above)
Comment (jdekonin) [reply buildenv/docker/mkdocker.sh]: In my latest push I cleaned up the ulimit required change for an added user, running as root ulimit value are unlimited so no compile issues are seen.  Also removed PS1 as it not added huge value.

File: buildenv/docker/mkdocker.sh (hunk 5 above)
Comment (jdekonin) [reply buildenv/docker/mkdocker.sh:554]: Commented out in my latest push.

File: buildenv/jenkins/docker-nodes/jenkinsfile-build-container.groovy (hunk 2 above)
Comment (jdekonin) [reply buildenv/jenkins/docker-nodes/jenkinsfile-build-container.groovy]: I decided against enabling cuda for all platforms here as the files extracted would be from ub16.  I doubt many will be wanting to use ub16 as a container as its out of support.  UB18+ and RH7+ we should likely aim for a version of cuda that would more closely align with the arch and os release.  I think that should be done under a different PR.

File: buildenv/jenkins/docker-nodes/jenkinsfile-build-container.groovy (hunk 3 above)
Comment (jdekonin) [reply buildenv/jenkins/docker-nodes/jenkinsfile-build-container.groovy]: Corrected to the location in the bugzilla with in my latest push.  I also fixed the REPOSITORY value for the same reason.

**Comment by jdekonin** [comment]
jenkins build docker x86 centos6

**Comment by jdekonin** [comment]
Looks like PR trigger for this job isn't working after the move.  I've modified the config for scheduled job to my repo and branch.
https://openj9-jenkins.osuosl.org/job/Build-Jenkins-Agent-Container/12/console

I believe this is ready for future review.  Once the job has completed, I still want to test to double check that it works for a compile as expected though.

**Comment by AdamBrousseau** [comment]
jenkins build docker x86 centos6

**Comment by jdekonin** [comment]
Updated the PR job config with the new location for the groovy file and try again.  The scheduled job works fine, but it would be nice to see the PR job work as well.  Latest push was a minor update to the readme.

jenkins build docker x86 centos6

**Comment by jdekonin** [comment]
jenkins build docker x86 centos6

File: buildenv/docker/mkdocker.sh [hunk 7]
```diff
@@ -671,7 +674,7 @@ print_dockerfile() {
 if [ $cuda != no ] ; then
   install_cuda
 fi
-  install_freemarker
+  #install_freemarker
```
Comment (keithc-ca) [inline buildenv/docker/mkdocker.sh:677]: I'd prefer to see this controlled by an option, say, `--freemarker`.

File: buildenv/docker/mkdocker.sh (hunk 4 above)
Comment (keithc-ca) [reply buildenv/docker/mkdocker.sh]: Why are we not heeding the warning in that readme file?
> The whole CentOS 6 is *dead* and *shouldn't* be used anywhere at *all*

File: buildenv/docker/mkdocker.sh (hunk 7 above)
Comment (jdekonin) [reply buildenv/docker/mkdocker.sh]: Added to this latest commit.

File: buildenv/jenkins/docker-nodes/README.md [hunk 8]
```diff
@@ -34,32 +34,32 @@ same environment every time.
...
 | Ubuntu18  |  Test Only   |   Test Only   |   Test Only   |
-| Centos6.9 | Build & Test | Not Available | Not Available |
+| Centos6   | Build & Test | Not Available | Not Available |
 | Centos7   | To be Added  |  To be Added  | Build & Test  |
 | Windows   | To be Added  | Not Available | Not Available |
 
 ## Eclipse OpenJ9 CI
 Generated Dockerfiles are used by two Jenkins jobs at Eclipse to
-build docker images. The first [job](https://ci.eclipse.org/openj9/view/Build/job/Build-Jenkins-Agent-Container/) is run manually
+build docker images. The first [job](/https://openj9-jenkins.osuosl.org/job/Build-Jenkins-Agent-Container/) is run manually
```
[the remark this answers is not in the document]
Comment (jdekonin) [reply buildenv/jenkins/docker-nodes/README.md]: Accidental, corrected in the latest commit.

File: buildenv/docker/mkdocker.sh [hunk 9]
```diff
@@ -532,32 +536,31 @@ prepare_user() {
...
   echo ""
   echo "# Add user home and copy authorized_keys and known_hosts."
-  echo "RUN useradd -ms /bin/bash $user \\"
+  echo "RUN useradd -ms /bin/bash --uid $userid $user \\"
   echo " && mkdir /home/$user/.ssh \\"
   echo " && chmod 700 /home/$user/.ssh"
   echo "COPY authorized_keys known_hosts /home/$user/.ssh/"
-  echo "RUN chown -R $user:$user /home/$user \\"
-  echo " && chmod 600 /home/$user/.ssh/authorized_keys /home/$user/.ssh/known_hosts"
+  echo "RUN chmod 600 /home/$user/.ssh/authorized_keys /home/$user/.ssh/known_hosts \\"
```
[the remark this answers is not in the document]
Comment (jdekonin) [reply buildenv/docker/mkdocker.sh]: Added to this latest commit.

File: buildenv/docker/mkdocker.sh (hunk 4 above)
Comment (jdekonin) [reply buildenv/docker/mkdocker.sh]: We are still using as this is the operating system that is used at Eclipse Adoptium for JDK8 x86(-64) support.

**Comment by jdekonin** [comment]
jenkins build docker x86 centos6

File: buildenv/docker/mkdocker.sh [hunk 10]
```diff
@@ -33,16 +33,17 @@ usage() {
   echo "  --arch=...     specify the processor architecture (default: host architecture)"
   echo "  --build        build the docker image (overrides '--print')"
   echo "  --cuda         include CUDA header files"
+  echo "  --freemarker   include freemarker.jar"
```
Comment (keithc-ca) [inline buildenv/docker/mkdocker.sh:36]: The options were listed in alphabetical order, and the variables and cases were ordered likewise.

File: buildenv/docker/mkdocker.sh [hunk 11]
```diff
@@ -223,10 +229,15 @@ fi
...
-if [ $version = 6.9 ] ; then
-  echo " && yum -y install https://repo.ius.io/ius-release-el6.rpm \\"
+if [ $version = 6 ] ; then
+  echo "RUN sed -i -e 's|mirrorlist|#mirrorlist|' \\"
+  echo "           -e 's|#baseurl=http://mirror.centos.org/centos/\$releasever|baseurl=https://vault.centos.org/6.10|' \\"
+  echo "           /etc/yum.repos.d/CentOS-Base.repo"
+  echo ""
 fi
+  echo "RUN yum -y update \\"
+  echo " && echo "assumeyes=1" >> /etc/yum.conf \\"
```
Comment (keithc-ca) [inline buildenv/docker/mkdocker.sh:239]: What is this for? Isn't it redundant with the `-y` option in `yum install -y`?

File: buildenv/docker/mkdocker.sh (hunk 10 above)
Comment (jdekonin) [reply buildenv/docker/mkdocker.sh]: Moved

File: buildenv/docker/mkdocker.sh (hunk 11 above)
Comment (jdekonin) [reply buildenv/docker/mkdocker.sh]: The extra line for yum.conf was because I had a failure/prompt with the epel-release repo line.  But I cannot seem to reproduce in a minimal testcase.  I removed.

File: buildenv/jenkins/docker-nodes/jenkinsfile-build-container.groovy (hunk 2 above)
Comment (keithc-ca) [reply buildenv/jenkins/docker-nodes/jenkinsfile-build-container.groovy]: It's only copying some header files from that image; only the version of CUDA should be relevant, but it should probably be updated to a supported OS version.

File: buildenv/docker/mkdocker.sh (hunk 10 above)
Comment (keithc-ca) [reply buildenv/docker/mkdocker.sh]: I wasn't going to insist, but thanks.

File: buildenv/docker/mkdocker.sh (hunk 4 above)
Comment (pshipton) [reply buildenv/docker/mkdocker.sh]: Not just jdk8, but jdk11+ as well.

**Comment by jdekonin** [comment]
jenkins build docker x86 centos6

**Review by keithc-ca — APPROVED** [review]

File: buildenv/jenkins/docker-nodes/jenkinsfile-build-container.groovy (hunk 2 above)
Comment (keithc-ca) [reply buildenv/jenkins/docker-nodes/jenkinsfile-build-container.groovy]: I expect we want to stick with version 9.0 for a while, which is only available on ubuntu 16.04; see https://hub.docker.com/r/nvidia/cuda/tags?page=1&ordering=last_updated&name=9.0.
